cpp/src/filling/repeat.cu

```diff
@@ -0,0 +1,150 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/copying.hpp>
+#include <cudf/filling.hpp>
+#include <cudf/types.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_factories.hpp>
+#if 0  // TODO: enable once the scatter/gather PR is merged
+#include <cudf/detail/gather.cuh>
+#endif
+#include <cudf/detail/repeat.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/mr/device_memory_resource.hpp>
+
+// for gdf_scalar, unnecessary once we switch to cudf::scalar
+#include <cudf/types.h>
+// for gdf_dtype_of, unnecessary once we switch to cudf::scalar
+#include <cudf/utilities/legacy/type_dispatcher.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/transform_output_iterator.h>
+#include <thrust/scan.h>
+#include <thrust/binary_search.h>
+
+#include <cuda_runtime.h>
+
+#include <memory>
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+std::unique_ptr<table> repeat(table_view const& input_table,
+                   column_view const& count, bool check_count,
+                   cudaStream_t stream,
+                   rmm::mr::device_memory_resource* mr) {
+  // TODO: can't this be of any integral type?
+  CUDF_EXPECTS(count.type().id() == type_to_id<size_type>(),
+               "count column should be of index type");
+  CUDF_EXPECTS(input_table.num_rows() == count.size(),
+               "in and count must have equal size");
+  CUDF_EXPECTS(count.has_nulls() == false, "count cannot contain nulls");
+
+  if (input_table.num_rows() == 0) {
+    return cudf::experimental::empty_like(input_table);
+  }
+  
+  rmm::device_vector<size_type> offsets(input_table.num_rows());
+  thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         count.begin<size_t>(), count.end<size_t>(), offsets.begin());
+  if (check_count == true) {
+    CUDF_EXPECTS(thrust::is_sorted(rmm::exec_policy(stream)->on(stream),
+                                   offsets.begin(), offsets.end()) == true,
+                 "count has negative values or the resulting table has more \
+                  rows than size_type's limit.");
+  }
+
+  auto output_size = size_type{offsets.back()};
+  auto p_indices = make_numeric_column(data_type{type_to_id<size_type>()},
```

Comment:
Does this need to be a column? Or can it just be a device vector?

Reply:
https://github.com/rapidsai/cudf/pull/3028/files#diff-f159bfeecf2a9d54754c4d6c6fd9bee4R44

The gather function takes a column_view as a gather map and AFAIK rmm::device_vector does not convert to a column_view... Am I missing other gather functions? (Maybe a version taking an iterator will be added?)

Reply:
> (Maybe a version taking an iterator will be added?)

Lucky you! It already exists!

https://github.com/rapidsai/cudf/pull/3028/files#diff-1511263c1493ba98239e38a46ebf1d8aR230